Mathlib/NumberTheory/ModularForms/EisensteinSeries/QExpansion.lean

```diff
@@ -0,0 +1,188 @@
+import Mathlib.Analysis.Complex.SummableUniformlyOn
+import Mathlib.Analysis.SpecialFunctions.Trigonometric.Cotangent
+
+/-!
+# Einstein series Q-expansions
+
+We give some identities for q-expansions of Eisenstein series that will be used in describing their
+Q-expansions.
+
+-/
+
+open Set Metric TopologicalSpace Function Filter Complex
+ EisensteinSeries
+
+open _root_.UpperHalfPlane hiding I
+
+open scoped Topology Real Nat Complex Pointwise
+
+local notation "ℍₒ" => upperHalfPlaneSet
+
+lemma iteratedDerivWithin_cexp_mul_const (k m : ℕ) (p : ℝ) {S : Set ℂ} (hs : IsOpen S) :
+    EqOn (iteratedDerivWithin k (fun s : ℂ ↦ cexp (2 * π * I * m * s / p)) S)
+    (fun s ↦ (2 * π * I * m / p) ^ k * cexp (2 * π * I * m * s / p)) S := by
+  apply EqOn.trans (iteratedDerivWithin_of_isOpen hs)
+  intro x hx
+  have : (fun s ↦ cexp (2 * π * I * m * s / p)) = (fun s ↦ cexp (((2 * π * I * m) / p) * s)) := by
+    ext z
+    ring_nf
+  simp only [this, iteratedDeriv_cexp_const_mul]
+  ring_nf
+
+private lemma aux_IsBigO_mul (k l : ℕ) (p : ℝ) {f : ℕ → ℂ}
+    (hf : f =O[atTop] (fun n ↦ ((n ^ l) : ℝ))) :
+    (fun n ↦ f n * (2 * π * I * n / p) ^ k) =O[atTop] (fun n ↦ (↑(n ^ (l + k)) : ℝ)) := by
+  have h0 : (fun n : ℕ ↦ (2 * π * I * n / p) ^ k) =O[atTop] (fun n ↦ ((n ^ k) : ℝ)) := by
+    have h1 : (fun n : ℕ ↦ (2 * π * I * n / p) ^ k) =
+      (fun n : ℕ ↦ ((2 * π * I / p) ^ k) * n ^ k) := by
+      ext z
+      ring
+    simpa [h1] using (Complex.isBigO_ofReal_right.mp (Asymptotics.isBigO_const_mul_self
+      ((2 * π * I / p) ^ k) (fun (n : ℕ) ↦ (↑(n ^ k) : ℝ)) atTop))
+  simp only [Nat.cast_pow] at *
+  convert hf.mul h0
+  ring
+
+open BoundedContinuousFunction in
+theorem summableLocallyUniformlyOn_iteratedDerivWithin_qExpansion (k l : ℕ) {f : ℕ → ℂ} {p : ℝ}
+    (hp : 0 < p) (hf : f =O[atTop] (fun n ↦ ((n ^ l) : ℝ))) :
+    SummableLocallyUniformlyOn (fun n ↦ (f n) •
+    iteratedDerivWithin k (fun z ↦ cexp (2 * π * I * z / p) ^ n) ℍₒ) ℍₒ := by
+  apply SummableLocallyUniformlyOn_of_locally_bounded isOpen_upperHalfPlaneSet
+  intro K hK hKc
+  haveI : CompactSpace K := isCompact_univ_iff.mp (isCompact_iff_isCompact_univ.mp hKc)
+  let c : ContinuousMap K ℂ := ⟨fun r : K ↦ Complex.exp (2 * π * I * r / p), by fun_prop⟩
+  let r : ℝ := ‖mkOfCompact c‖
+  have hr : ‖r‖ < 1 := by
+    simp only [norm_norm, r, norm_lt_iff_of_compact Real.zero_lt_one, mkOfCompact_apply,
+      ContinuousMap.coe_mk, c]
+    intro x
+    have h1 : cexp (2 * π * I * (x / p)) = cexp (2 * π * I * x / p) := by
+      ring_nf
+    simpa using h1 ▸ UpperHalfPlane.norm_exp_two_pi_I_lt_one ⟨((x : ℂ) / p) , by aesop⟩
+  refine ⟨_, by simpa using (summable_norm_mul_geometric_of_norm_lt_one' hr
+    (Asymptotics.isBigO_norm_left.mpr (aux_IsBigO_mul k l p hf))), ?_⟩
+  intro n z hz
+  have h0 := pow_le_pow_left₀ (by apply norm_nonneg _) (norm_coe_le_norm (mkOfCompact c) ⟨z, hz⟩) n
+  simp only [norm_mkOfCompact, mkOfCompact_apply, ContinuousMap.coe_mk, ←
+    exp_nsmul', Pi.smul_apply,
```

Comment:
```suggestion
  simp only [norm_mkOfCompact, mkOfCompact_apply, ContinuousMap.coe_mk, ← exp_nsmul', Pi.smul_apply,
```
if it fits on the line; in any case, don't leave a dangling `←` at the end of the line.